File: 1/Hazi.java
import java.util.*;
import java.io.*;

public class Hazi {
	HashMap<String,Integer> hn;
	HashMap<String,String> cameFrom;
	class Node implements Comparable<Node> {
		String name;
		int f, g, h;
		public Node(String name, int g, int h) {
			this.name = name;
			this.f = g + h;
			this.g = g;
			this.h = h;
		}
		public int compareTo(Node n) {
			Integer fi = new Integer(f);
			Integer nfi = new Integer(n.f);
			return fi.compareTo(nfi);
		}
	}
	boolean nameInList(String y, List l) {
		for (Iterator i = l.listIterator(); i.hasNext();) {
			Node n = (Node) i.next();
			if (n.name.equals(y))
				return true;
		}
		return false;
	}
	Node nodeFromList(String y, List l) {
		for (Iterator i = l.listIterator(); i.hasNext();) {
			Node n = (Node) i.next();
			if (n.name.equals(y))
				return n;
		}
		return null;
	}
	List<String> reconstructPath(HashMap<String,String>cameFrom, String currentNode){
		boolean in_keys = false;
		for (Iterator i = cameFrom.keySet().iterator(); i.hasNext(); ) {
			String s = (String) i.next();
			if (s.equals(currentNode)) {
				in_keys = true;
				break;
			}
		}
		if (in_keys) {
			List<String> p = reconstructPath(cameFrom, cameFrom.get(currentNode));
			p.add(currentNode);
			return p;
		} else {
			List<String> p = new LinkedList<String>();
			p.add(currentNode);
			return p;
		}
	}
	boolean aStar(String start, String end) {
		try {
			BufferedWriter sock = new BufferedWriter(
					new OutputStreamWriter(
						new FileOutputStream("output.txt"), "8859_2")
					);
			List<Node> openlist = new LinkedList<Node>();
			openlist.add(new Node(start, 0, hn.get(start)));
			List<Node> closedlist = new LinkedList<Node>();
			int count = 0;
			while (openlist.size() > 0) {
				Collections.sort(openlist);
				sock.write("(:openlist " + count);
				for (Iterator i = openlist.listIterator(); i.hasNext();) {
					Node n = (Node) i.next();
					sock.write(" ("+n.f+" "+n.name+")");
				}
				sock.write(")");
				sock.newLine();
				sock.write("(:closedlist " + count);
				for (Iterator i = closedlist.listIterator(); i.hasNext();) {
					Node n = (Node) i.next();
					sock.write(" ("+n.f+" "+n.name+")");
				}
				sock.write(")");
				sock.newLine();
				
				Node x = openlist.get(0);
				openlist.remove(0);
				if (x.name.equals(end)) {
					sock.write("(:sol "+x.f);
					List<String> l = reconstructPath(cameFrom, end);
					for (Iterator i = l.listIterator(); i.hasNext();) {
						String s = (String) i.next();
						sock.write(" "+s);
					}
					sock.write(")");
					sock.newLine();
					return true;
				}
				closedlist.add(x);
			}
		} catch(Exception e) {
			e.printStackTrace();
		}
		return false;
	}
	public Hazi() {
	}
}

File: a-star/hazi.py
#!/usr/bin/env python

import sys, codecs

class Node:
	def __init__(self, name, g, h):
		self.name = name
		self.f = g + h
		self.g = g
		self.h = h

def sort_node(a, b):
	return cmp(a.f, b.f)

def name_in_list(y, l):
	for i in l:
		if y == i.name:
			return True
	return False

def node_from_list(y, l):
	for i in l:
		if y == i.name:
			return i

def reconstruct_path(came_from,current_node):
	if current_node in came_from.keys():
		p = reconstruct_path(came_from,came_from[current_node])
		return p + [current_node]
	else:
		return [current_node]

def a_star(start, end):
	sock = codecs.open("output.txt", "w", "ISO-8859-2")
	openlist = [Node(start, 0, hn[start])]
	closedlist = []
	count = 0
	while len(openlist):
		openlist.sort(cmp=sort_node)
		sock.write("(:openlist %s" % count)
		for i in openlist:
			sock.write(" (%s %s)" % (i.f, i.name))
		sock.write(")\n")
		sock.write("(:closedlist %s" % count)
		for i in closedlist:
			sock.write(" (%s %s)" % (i.f, i.name))
		sock.write(")\n")

		x = openlist.pop(0)
		if x.name == end:
			sock.write("(:sol %s " % x.f)
			sock.write(" ".join(reconstruct_path(came_from,end)))
			sock.write(")\n")
			return True
		closedlist.append(x)
		for y in gn[x.name].keys():
			if name_in_list(y, closedlist):
				continue
			tentative_g_score = x.g + gn[x.name][y]
			tentative_is_better = False
			if not name_in_list(y, openlist):
				openlist.append(Node(y, tentative_g_score, hn[y]))
				tentative_is_better = True
			elif tentative_g_score < node_from_list(y, openlist).g:
				tentative_is_better = True
			if tentative_is_better == True:
				came_from[y] = x.name
		count += 1
	return False

inhn = False
ingn = False
start = None
end = None
hn = {}
gn = {}
came_from = {}
sock = codecs.open(sys.argv[1], "r", "ISO-8859-2")
for i in sock.readlines():
	line = i.strip()
	if line.startswith("(:start"):
		start = line[8:-1]
	elif line.startswith("(:end"):
		end = line[6:-1]
	elif line.startswith("(:hn"):
		inhn = True
	elif line.startswith("(:gn"):
		ingn = True
	elif line.startswith("("):
		if inhn:
			items = line[1:-1].split(' ')
			hn[items[0]] = int(items[1])
		elif ingn:
			items = line[1:-1].split(' ')
			if items[0] not in gn.keys():
				gn[items[0]] = {}
			gn[items[0]][items[1]] = int(items[2])
	elif line.startswith(")"):
		if inhn:
			inhn = False
		elif ingn:
			ingn = False
sock.close()
a_star(start, end)
